src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs

```diff
@@ -220,13 +215,27 @@ private void SetupSendBuffers(ReadableBuffer buffer)
                         _output.Advance(buffer.End);
                     }
                 }
-
-                // We're done reading
-                _output.Complete();
             }
             catch (Exception ex)
             {
-                _output.Complete(ex);
+                if (ex is SocketException se && se.SocketErrorCode == SocketError.OperationAborted || ex is ObjectDisposedException)
+                {
+                    // We had to abort the write
+                    error = null;
+                }
+                else if (ex is IOException)
+                {
+                    error = ex;
+                }
+                else
+                {
+                    error = new IOException(ex.Message, ex);
+                }
+            }
+            finally
+            {
+                _connectionContext.OnConnectionClosed(error);
```

Comment:
Semantically it makes sense to call this after the socket is disposed. If it socket is always closed after the call to OnApplicationComplete(), OnConnectionClosed() wouldn't need to exist.